src/transactions/RestoreFootprintOpFrame.cpp

```diff
@@ -47,180 +54,346 @@ RestoreFootprintOpFrame::RestoreFootprintOpFrame(
         }
 
-        metrics.mLedgerReadByte += entrySize;
-        if (resources.diskReadBytes < metrics.mLedgerReadByte)
+        int64_t rentFee = rust_bridge::compute_rent_fee(
+            mAppConfig.CURRENT_LEDGER_PROTOCOL_VERSION, ledgerVersion,
+            rustEntryRentChanges,
+            mSorobanConfig.rustBridgeRentFeeConfiguration(), ledgerSeq);
+        if (!mRefundableFeeTracker->consumeRefundableSorobanResources(
+                0, rentFee, getLedgerVersion(), mSorobanConfig, mAppConfig,
+                mOpFrame.mParentTx, mDiagnosticEvents))
         {
-            diagnosticEvents.pushError(
-                SCE_BUDGET, SCEC_EXCEEDED_LIMIT,
-                "operation byte-read resources exceeds amount specified",
-                {makeU64SCVal(metrics.mLedgerReadByte),
-                 makeU64SCVal(resources.diskReadBytes)});
-            innerResult(res).code(RESTORE_FOOTPRINT_RESOURCE_LIMIT_EXCEEDED);
-            return {false, {}};
+            innerResult(mRes).code(
+                RESTORE_FOOTPRINT_INSUFFICIENT_REFUNDABLE_FEE);
+            return false;
         }
+        innerResult(mRes).code(RESTORE_FOOTPRINT_SUCCESS);
+        return true;
+    }
+};
 
-        // To maintain consistency with InvokeHostFunction, TTLEntry
-        // writes come out of refundable fee, so only add entrySize
-        metrics.mLedgerWriteByte += entrySize;
-        if (!validateContractLedgerEntry(lk, entrySize, sorobanConfig,
-                                         appConfig, mParentTx,
-                                         diagnosticEvents))
+class RestoreFootprintPreV23ApplyHelper
+    : virtual public RestoreFootprintApplyHelper,
+      virtual public PreV23LedgerAccessHelper
+{
+  public:
+    RestoreFootprintPreV23ApplyHelper(
+        AppConnector& app, AbstractLedgerTxn& ltx, OperationResult& res,
+        std::optional<RefundableFeeTracker>& refundableFeeTracker,
+        OperationMetaBuilder& opMeta, RestoreFootprintOpFrame const& opFrame)
+        : RestoreFootprintApplyHelper(app, res, refundableFeeTracker, opMeta,
+                                      opFrame)
+        , PreV23LedgerAccessHelper(ltx)
+    {
+    }
+
+    std::optional<LedgerEntry>
+    getHotArchiveEntry(LedgerKey const& key) override
+    {
+        // There is no hot archive pre-23.
+        return std::nullopt;
+    }
+    bool
+    entryWasRestored(LedgerKey const& key) override
+    {
+        // NB: even though pre-23 can answer this question precisely, the code
+        // in pre-23 wasn't sensitive to it, so we preserve that functionality.
+        return false;
+    }
+    void
+    restoreEntry(LedgerKey const& lk, LedgerEntry const& entry_,
+                 LedgerKey const& ttlKey,
+                 std::optional<LedgerEntry> const& ttlEntryOpt,
+                 uint32_t restoredLiveUntilLedger) override
+    {
+        // Entry exists in the live BucketList if we get to this point due to
```

Comment:
Nit: This copy paste comment doesn't make sense anymore. Change to "If we restore an entry from pre 23, we know it must exist in the LiveBucketList since we don't support the Hot Archive yet.

Reply:
fixed.